auth/auth.go

```diff
@@ -477,24 +477,51 @@ func (auth *Authenticator) AuthenticateUntrustedJWT(token string, providers OIDC
 }
 
 // Authenticates a user based on a JWT token obtained directly from a provider (auth code flow, refresh flow).
-// Verifies the token claims, but doesn't require signature verification.
+// Verifies the token claims, but doesn't require signature verification if allow_unsigned_provider_tokens is enabled.
 // If the token is validated but the user for the username defined in the subject claim doesn't exist,
 // creates the user when autoRegister=true.
-func (auth *Authenticator) AuthenticateTrustedJWT(token string, provider *OIDCProvider) (user User, tokenExpiry time.Time, err error) {
+func (auth *Authenticator) AuthenticateTrustedJWT(token string, provider *OIDCProvider, callbackURLFunc OIDCCallbackURLFunc) (user User,
+	tokenExpiry time.Time, err error) {
 	base.Debugf(base.KeyAuth, "AuthenticateTrustedJWT called with token: %s", base.UD(token))
 
-	// Verify claims - ensures that the token we received from the provider is valid for Sync Gateway
-	identity, err := VerifyClaims(token, provider.ClientID, provider.Issuer)
-	if err != nil {
-		base.Debugf(base.KeyAuth, "Error verifying raw token in AuthenticateTrustedJWT: %v", err)
-		return nil, time.Time{}, err
+	var identity *Identity
+	if provider.AllowUnsignedProviderTokens {
+		// Verify claims - ensures that the token we received from the provider is valid for Sync Gateway
+		identity, err = VerifyClaims(token, provider.ClientID, provider.Issuer)
+		if err != nil {
+			base.Debugf(base.KeyAuth, "Error verifying raw token in AuthenticateTrustedJWT: %v", err)
+			return nil, time.Time{}, err
+		}
+	} else {
+		// Get client for issuer
```

Comment:
This code looks identical the code used in AuthenticateUntrustedJWT.  Any reason we're not pulling the common code into a shared function?

Reply:
Initially, I was about to encapsulate the claims and signature verification into a separate function and reuse that in both AuthenticateUntrustedJWT and AuthenticateTrustedJWT. But later decided to not touch AuthenticateUntrustedJWT to make this PR specific to AuthenticateUntrustedJWT and make this PR lighter. But it is reasonable to make do that change. Included it in the latest revision.